test/distributed/test_c10d_common.py

```diff
@@ -1326,6 +1335,208 @@ def test_send_recv(self):
+                        {},
+                        name="wait_comm"
+                    )
+                    t.proxy = proxy_res
+                    #proxy_res.node.meta['tensor_meta'] = _extract_tensor_metadata(t)
+                    return t
+                else:
+                    # in eager mode, simply wait
+                    w.wait()
+                    return t
+            else:
+                return e
+
+        args = tree_map(unwrap, args)
+        kwargs = tree_map(unwrap, kwargs)
+
+        return func(*args, **kwargs)
+
+
+def wrap_comm_result(result):
+    return ([CommResultTensor(result[0][0], result[1])], result[1])
+
+
+class CommTensor(torch.Tensor):
+    @staticmethod
+    def __new__(cls, tensor: torch.Tensor):
+
+        r = torch.Tensor._make_wrapper_subclass(  # type: ignore[attr-defined]
+            cls,
+            tensor.size(),
+            strides=list(tensor.stride()),
+            dtype=tensor.dtype,
+            device=tensor.device,
+            layout=tensor.layout,
+            requires_grad=tensor.requires_grad,
+        )
+        r._tensor = tensor
+        return r
+
+    def __repr__(self):
+        return f"CommTensor({self._tensor})"
+
+    # disable __torch_function__ so that CommTensor can recursively dispatch
+    # ProxyTensor in make_fx
+    __torch_function__ = _disabled_torch_function_impl
+
+    @classmethod
+    def __torch_dispatch__(cls, func, types, args=(), kwargs=None):
+
+        tracing = False
+        proxy_mode = None
+        def unwrap(e):
+            if isinstance(e, CommTensor):
+                nonlocal tracing, proxy_mode
+                if isinstance(e._tensor, ProxyTensor):
+                    tracing = True
+                    proxy_mode = e._tensor.proxy_mode
+                return e._tensor
+            else:
+                return e
+
+        def unwrap_proxy(e):
+            return e.proxy if isinstance(e, ProxyTensor) else e
+
+        def unwrap_elem(e):
+            return e.elem if isinstance(e, ProxyTensor) else e
+
+        args = tree_map(unwrap, args)
+        kwargs = tree_map(unwrap, kwargs)
+
+        if "allreduce_" in func.__name__:
+            if tracing:
+                proxy_args = tree_map(unwrap_proxy, args)
+                proxy_kwargs = tree_map(unwrap_proxy, kwargs)
+
+                proxy_res = func(*proxy_args, **proxy_kwargs)
+                comm_tensor_proxy = proxy_mode.tracer.create_proxy(
+                    'call_function',
+                    wrap_comm_result,
+                    (proxy_res, ),
+                    {},
+                    name="comm_result"
+                )
+
+                with no_dispatch():
+                    elem_args = tree_map(unwrap_elem, args)
+                    elem_kwargs = tree_map(unwrap_elem, kwargs)
+
+                    out = func(*elem_args, **elem_kwargs)
+
+                return wrap_output(out, comm_tensor_proxy, constant=None, proxy_mode=proxy_mode)
```

Comment:
Hey @Chillee, below is the `print_tabular` result. Do you know why `aten.mul.Tensor` is still using `add_tensor` instead of `getitem`? 

```
opcode         name                target                                         args                                      kwargs
-------------  ------------------  ---------------------------------------------  ----------------------------------------  --------
placeholder    x_1                 x_1                                            ()                                        {}
call_function  add_tensor          aten.add.Tensor                                (x_1, x_1)                                {}
get_attr       _tensor_constant0   _tensor_constant0                              ()                                        {}
call_function  allreduce__default  c10d.allreduce_.default                        ([add_tensor], _tensor_constant0, 0, -1)  {}
call_function  comm_result         <function wrap_comm_result at 0x7f487f58a820>  (allreduce__default,)                     {}
call_function  getitem             <built-in function getitem>                    (comm_result, 0)                          {}
call_method    call                __call__                                       (getitem, 0)                              {}
call_function  getitem_1           <built-in function getitem>                    (comm_result, 1)                          {}
call_function  mul_tensor          aten.mul.Tensor                                (add_tensor, 2)                           {}
get_attr       _tensor_constant1   _tensor_constant1                              ()                                        {}
output         output              output                                         (_tensor_constant1,)                      {}
``` 

For context, I cannot directly call `func(*args, **kwargs)` for `allreduce_` and operate on the output, because at that time the returned `work` handle no longer has an associated proxy and will be treated as a constant. I believe I have to somehow capture the runtime `work` handle in the graph? So I wanted to directly operate on this [`proxy_res`](https://github.com/pytorch/pytorch/blob/8a6b076196f4169910b2c98d37e1faeaec316050/torch/fx/experimental/proxy_tensor.py#L168) where I still have a chance to use `getitem` to retrieve runtime work handle.

Reply:
oh I need to update `.proxy` for the input arg as it's inplace